Fix infinite request bug

squeevee opened the pull request:
(due to git clumsiness, this PR includes the entirety of my other PR, so probably handle that one first)

So, there was a bug that caused requests to be spawned infinitely. say the route http://example.com/foo is mapped to be handled by the GetController but isn't the id of any object in the database. Server receives request on http://example.com/foo, calls GetController::handle(), which calls ObjectService::dereference(), which fails to find the object in the database and decides to look it up on the "remote" server. Except it's not the remote server, it's the same server, so the same stack gets called again, and so on.

Simple fix, dereference() now also accepts a bool localOnly (default false) that can tell it to only look in the database. GetController and PostController call it with true.

Also a smidge of refactoring for convenience.

squeevee committed:
minor changes for operability with calling app, see PR

squeevee committed:
merge

squeevee committed:
Another quick syntax fix

squeevee committed:
Fix infinite request bug

squeevee committed:
Revert "Fix infinite request bug", meant to put that in a separate branch

This reverts commit a5b75cd7e95b9d92a2a12594aafe35296156f44b.

jdormit commented:
Good catch! I'll hold off merging this until we merge #3 to avoid the git history from getting too gnarly.

jdormit commented:
I've thought about this a bit, and I think it would be a cleaner solution to check in the `dereference` function to see if the object being dereferenced is local (has the same hostname in its id as `$request->getHost()`. If so, we just won't make a network call, returning `null` immediately if the object isn't in the DB.

squeevee commented:
Yeah, you're right that's cleaner. I'll put that into this branch

squeevee committed:
branch sync

squeevee committed:
missed some details

squeevee committed:
manually reapplying changes because i can't figure out git rebase

squeevee commented:
Out of curiosity, why does it remove the queries from the urls to check them against ID's?

jdormit commented:
The query params are removed because I (and other implementations) am using query params to represent page offsets in collections, so e.g. the collection id would be `https://example.com/collection/1` and the first page would be `https://example.com/collection/1?offset=1`. Both of those URIs should still dereference to the same object in the graph, the collection with id `https://example.com/collection/1`.

eleanorhawk committed:
Merge remote-tracking branch 'upstream/master'

eleanorhawk committed:
Merge branch 'master' into squeevee-dev-1

eleanorhawk committed:
merge stash

squeevee committed:
Removed metadata mappings

squeevee committed:
cleanup

squeevee commented:
New changes that are here even though I meant them for #3 (sorry).
- Added composer.lock to .gitignore and deleted it so it will no longer be tracked. (Since it's local configuration state)
- Removed my metadata mapping stuff
- Removed the inline metadata mapping in ActivityPubObject.php, Field.php, and PrivateKey.php so now table prefixes work.

And now `ObjectService::dereference()` compares the host of `$id` to the incoming request to automatically determine if the object is local.

jdormit commented:
Sorry for the delay here - I'm going to try to find some time to review this in the next couple of weeks.

squeevee commented:
> Sorry for the delay here - I'm going to try to find some time to review this in the next couple of weeks.

no worries.
